keps/sig-api-machinery/3157-watch-list/README.md

```diff
@@ -581,6 +582,73 @@ Then on the server side we:
 3. reject the request if waitUntilFreshAndBlock times out, thus forcing informers to retry.
 4. otherwise, construct the final list and send back to a client.
 
+### Replacing standard List request with WatchList mechanism for client-go's List method.
+
+Replacing the underlying implementation of the List method for client-go based clients (like typed or dynamic client) 
+with the WatchList mechanism requires ensuring that the data returned by both the standard List request and 
+the new WatchList mechanism remains identical. The challenge is that WatchList no longer retrieves the entire 
+list from the server at once but only receives individual items, which forces us to "manually" reconstruct 
+the list object on the client side.
+
+To correctly construct the list object on the client side, we need ListKind information.
+However, simply reconstructing the list object based on these data is not enough. 
+In the case of a standard List request, the server's response (a versioned list) is processed through a chain of decoders, 
+which can potentially modify the resulting list object. 
+A good example is the WithoutVersionDecoder, which removes the GVK information from the list object.
+Thus the "manually" constructed list object may not be consistent 
+with the transformations applied by the decoders, leading to differences.
+
+To ensure full compatibility, the server must provide a versioned empty list in the format requested by the client (e.g., protobuf representation). 
+We don't know how the client's decoder behaves for different encodings, i.e., whether the decoder actually supports 
+the encoding we intend to use for reconstruction. Therefore, to ensure maximal compatibility, we will ensure that 
+the encoding used for the reconstruction of the list matches the format that the client originally requested. 
+This guarantees that the returned list object can be correctly decoded by the client, 
+preserving the actual encoding format as intended.
+
+The proposed solution is to add a new annotation (`k8s.io/initial-events-list-blueprint`) to the object returned 
+in the bookmark event (The bookmark event is sent when the state is synced and marks the end of WatchList stream). 
+This annotation will store an empty, versioned list encoded as a Base64 string. 
+This annotation will be added to the same object/place the `k8s.io/initial-events-end` annotation is added.
+
+When the client receives such a bookmark, it will base64 decode the empty list and pass it to the decoder chain. 
```

Comment:
To be clear, clients should pass the decoded base64 text to precisely the same decoder that it uses to decode the objects embedded in watch events? Out of curiosity, what would break if the bookmark event's embedded object was itself an empty instance of the list type?

Reply:
>To be clear, clients should pass the decoded base64 text to precisely the same decoder that it uses to decode the objects embedded in watch events?

yes

> Out of curiosity, what would break if the bookmark event's embedded object was itself an empty instance of the list type?

changing the bookmark event's embedded object would break informer and all clients that didn't expect to find a different object in the event :(